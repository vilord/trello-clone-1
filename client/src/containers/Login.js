import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { loginUser } from '../actions/user';
import { resetUiError } from '../actions/ui';
import { Link } from 'react-router-dom';
import { Form, Button, Message, Loader, Dimmer } from 'semantic-ui-react';
import * as errors from '../constants/errors';

import gLogo from '../assets/G-logo.svg';
import './Login.css';

// TODO: Add sending Spinner
export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
    };

    // Bidings
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleChange(e, { name, value }) {
    const { error, resetError } = this.props;
    if (error) {
      resetError();
    }
    this.setState({ [name]: value });
  }

  handleSubmit(event) {
    event.preventDefault();
    const { loginUser, history } = this.props;
    const userCredentials = { ...this.state };
    loginUser(userCredentials, history);
  }

  render() {
    const { email, password } = this.state;
    const { error, header, message, serverError, sending } = this.props;
    const loginDisabled = !email || !password || sending;
    const host =
      process.env.NODE_ENV === 'development' ? 'http://localhost:3001' : '';
    const authURL = `${host}/auth/google`;

    return (
      <div className="Login">
        <h1>Login in to Trello</h1>
        <p>
          or <Link to="/signup">create an account</Link>
        </p>
        <Form
          className="Login-form"
          size="big"
          onSubmit={this.handleSubmit}
          error={error || serverError}
        >
          <Form.Input
            autoFocus
            type="text"
            placeholder="e.g., hermioneg@example.com"
            error={error}
            label={
              <label className="emailLabel">
                Email <span>(or username)</span>
              </label>
            }
            name="email"
            value={email}
            onChange={this.handleChange}
            disabled={sending}
          />
          <Form.Input
            label="Password"
            type="password"
            name="password"
            value={password}
            placeholder="e.g., ••••••••••"
            onChange={this.handleChange}
            error={error}
            disabled={sending}
          />
          <Message
            error
            header={header || 'Header'}
            content={message || 'Message'}
          />
          <Button
            className="Signup-button"
            size="large"
            type="submit"
            fluid
            primary={!loginDisabled}
            disabled={loginDisabled}
          >
            Log In
          </Button>
        </Form>
        <a className="Login-google" href={authURL}>
          <img src={gLogo} alt="" />
          <p>Log in with Google</p>
        </a>
        <Dimmer active={sending} inverted>
          <Loader size="large" inverted content="Loading" />
        </Dimmer>
      </div>
    );
  }
}

Login.defaultProps = {
  error: false,
  serverError: false,
  header: '',
  message: '',
  sending: false,
};

Login.propTypes = {
  error: PropTypes.bool,
  serverError: PropTypes.bool,
  header: PropTypes.string,
  message: PropTypes.string,
  sending: PropTypes.bool,
  loginUser: PropTypes.func.isRequired,
  resetError: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  const { kind, header, message } = state.ui.error;
  const error = kind === errors.wrongCredentials.kind;
  const serverError = kind === errors.serverError.kind;
  return {
    error,
    serverError,
    header,
    message,
    sending: state.ui.fetching.login,
  };
};

const mapDispatchToProps = dispatch => ({
  loginUser: (user, history) => dispatch(loginUser(user, history)),
  resetError: () => dispatch(resetUiError()),
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
