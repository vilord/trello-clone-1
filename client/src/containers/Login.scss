.Login {
  display: flex;
  flex-direction: column;

  width: 350px;

  h1 {
    font-size: 3em;

    margin-bottom: 0;
  }

  p {
    font-size: 16px;
  }
}

.Login-form {
  margin-top: 20px;
}

.emailLabel {
  span {
    color: #999;
    font-weight: 500;
  }
}

.Login-google {
  font-family: 'Roboto', sans-serif;

  display: flex;
  align-items: center;
  justify-content: center;

  width: 100%;
  height: 46px;
  margin-top: 20px;

  color: grey;
  border: 2px solid lightgrey;
  border-radius: 5px;

  img {
    height: 70%;
    margin-right: 20px;
  }

  &:hover {
    color: grey;
  }

  &:active {
    background-color: lightgray;
  }
}
