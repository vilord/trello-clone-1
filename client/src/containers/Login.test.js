import React from 'react';
import { shallow } from 'enzyme';
import { Login } from './Login';
import { Form } from 'semantic-ui-react';

describe('Login Component', () => {
  let component;
  const minProps = {
    loginUser: jest.fn(),
    resetError: jest.fn(),
  };

  beforeAll(() => {
    console.error = error => {
      throw new Error(error);
    };
  });

  beforeEach(() => {
    component = shallow(<Login {...minProps} />);
  });

  it('renders', () => {
    expect(component).toHaveLength(1);
  });

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('has className Login', () => {
    expect(component.hasClass('Login')).toBe(true);
  });

  it('sets state value to input fields', () => {
    const state = {
      email: 'johndoe@example.com',
      password: '123abc',
    };
    component.setState(state);
    expect(
      component
        .find(Form.Input)
        .first()
        .prop('value'),
    ).toBe(state.email);
    expect(
      component
        .find(Form.Input)
        .last()
        .prop('value'),
    ).toBe(state.password);
  });

  it('disables Button when either input is empty', () => {
    expect(component.find('Button').prop('disabled')).toBe(true);
  });

  it('enables Button when inputs have values', () => {
    component.setState({
      email: 'johndoe@example.com',
      password: '123abc',
    });
    expect(component.find('Button').prop('disabled')).toBe(false);
  });

  it('disables Button when sending', () => {
    component.setState({
      email: 'johndoe@example.com',
      password: '123abc',
    });
    component.setProps({ sending: true });
    expect(component.find('Button').prop('disabled')).toBe(true);
  });

  it('sends login when form is submitted', () => {
    component.setState({
      email: 'johndoe@example.com',
      password: '123abc',
    });

    const event = { preventDefault: jest.fn() };
    expect(minProps.loginUser).not.toHaveBeenCalled();

    component.find(Form).simulate('submit', event);
    expect(minProps.loginUser).toHaveBeenCalled();
  });

  it('resets UI errors on input change', () => {
    component.setProps({ error: true });
    expect(minProps.resetError).not.toHaveBeenCalled();

    const newEmail = {
      name: 'email',
      value: 'johndoe@example.com',
    };
    component
      .find(Form.Input)
      .first()
      .simulate('change', null, newEmail);
    expect(minProps.resetError).toHaveBeenCalled();

    const newPassword = {
      name: 'password',
      value: '123abc',
    };
    component
      .find(Form.Input)
      .last()
      .simulate('change', null, newPassword);
    expect(minProps.resetError).toHaveBeenCalledTimes(2);
  });
});
