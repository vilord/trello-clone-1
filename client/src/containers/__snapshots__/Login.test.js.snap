// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches snapshot 1`] = `
<div
  className="Login"
>
  <h1>
    Login in to Trello
  </h1>
  <p>
    or 
    <Link
      replace={false}
      to="/signup"
    >
      create an account
    </Link>
  </p>
  <Form
    as="form"
    className="Login-form"
    error={false}
    onSubmit={[Function]}
    size="big"
  >
    <FormInput
      as={[Function]}
      autoFocus={true}
      control={[Function]}
      disabled={false}
      error={false}
      label={
        <label
          className="emailLabel"
        >
          Email 
          <span>
            (or username)
          </span>
        </label>
      }
      name="email"
      onChange={[Function]}
      placeholder="e.g., hermioneg@example.com"
      type="text"
      value=""
    />
    <FormInput
      as={[Function]}
      control={[Function]}
      disabled={false}
      error={false}
      label="Password"
      name="password"
      onChange={[Function]}
      placeholder="e.g., ••••••••••"
      type="password"
      value=""
    />
    <Message
      content="Message"
      error={true}
      header="Header"
    />
    <Button
      as="button"
      className="Signup-button"
      disabled={true}
      fluid={true}
      primary={false}
      size="large"
      type="submit"
    >
      Log In
    </Button>
  </Form>
  <a
    className="Login-google"
    href="/auth/google"
  >
    <img
      alt=""
      src="G-logo.svg"
    />
    <p>
      Log in with Google
    </p>
  </a>
  <Dimmer
    active={false}
    inverted={true}
  >
    <Loader
      content="Loading"
      inverted={true}
      size="large"
    />
  </Dimmer>
</div>
`;
